### ISSUE TITLE
Rule proposal: Prevent use of extended native objects

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#3426]
Similar to [`no-extend-native`](http://eslint.org/docs/rules/no-extend-native), but the intention of this is to prevent usage of extended native objects, not creation.

When using external Node modules you can't control whether they extend native objects. I would still like to prevent usage of it. When the type can be detected it should check whether the property/method is a known native one, and if not, throw a violation.

A good example is the [colors](https://github.com/marak/colors.js/) module which extends the string prototype with [coloring setter properties](https://github.com/Marak/colors.js/blob/90dd85b173cc48d668412a84306bf6b01751dd21/lib/extendStringPrototype.js).

This should be a violation as there's is no `green` property on the native string prototype:

```js
var colors = require('colors');
console.log('unicorn'.green);
```

-

I've made a [module](https://github.com/sindresorhus/proto-props) with a static list of prototype properties for each type, which could potentially be used.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
This sounds like a good idea for a custom rule, but not something I'd like to address as part of the core.

**Comment by sindresorhus** [comment]
Too bad. I think it would be a lot more useful than `no-native-extend` as it's much more common to see usage of extended native objects than construction (that usually happens in libs). I guess I'll just have to show its usefulness through plugin popularity ;)

I've opened a module request. Happy to discuss there with any potential takers.

https://github.com/sindresorhus/module-requests/issues/25

**Comment by BYK** [comment]
@sindresorhus - just wondering, why not use TypeScript or flow at that point?

**Comment by sindresorhus** [comment]
I don't understand the question. ESLint using Flow would make this kind of check a lot easier to make, but I don't think Flow has any linting capabilities. No idea how it would even relate to TypeScript.

**Comment by BYK** [comment]
@sindresorhus - Oh, I thought you were trying to prohibit calls to undefined methods but now I realize you want to forbid the calls even if some framework defines it and ESLint misses that. Am I right?

**Comment by sindresorhus** [comment]
Yes. It should look at the usage, not the definition.

**Comment by sindresorhus** [comment]
@dustinspecker made a plugin https://github.com/dustinspecker/eslint-plugin-no-use-extend-native Yay!
